ensureCacheSpaceWithLock panics with variable sized keys

If variable sized keys are being used, then ensureCacheSpaceWithLock() can panic during the cache cleanup routine which loops through safe(). This occurs if the key being inserted is larger than the last key removed.

It would be better if either this key were inserted anyway (and thus exceeding the cache threshold) or the routine was modified so that it cleared some percent of the total cache size before proceeding.

On further review, I'm not convinced that WinstonPrivacy's fixes actually do anything, but I'm also not seeing any way for the cache size to get out of sync.

When Write is called, any existing value is blown out of the cache, with the cache size decremented by whatever was actually in the cache map.

Whenever we read an item and add it to the cache, it increments the cacheSize by whatever the size of the actual value may be.

I don't see a way in which we could, say, push a 100 byte value into the cache but increment the cacheSize by 200.

I've fooled around with tests to try and replicate the crash, but I cannot. I don't know how this happened:

```
panic: 833478 bytes still won't fit in the cache! (max 1048576 bytes)

goroutine 603606 [running]:
github.com/peterbourgon/diskv.(*Diskv).ensureCacheSpaceWithLock(0xc0001413b0,
0xcb7c6, 0x1eb2580, 0xc000d09588)
        github.com/peterbourgon/diskv@v2.0.1+incompatible/diskv.go:611 +0x2bd
github.com/peterbourgon/diskv.(*Diskv).cacheWithLock(0xc0001413b0,
0xc001933b60, 0x26, 0xc0170ae000, 0xcb7c6, 0xf9f82, 0xc000074060,
0x1001)
        github.com/peterbourgon/diskv@v2.0.1+incompatible/diskv.go:532 +0x3f
github.com/peterbourgon/diskv.(*Diskv).cacheWithoutLock(0xc0001413b0,
0xc001933b60, 0x26, 0xc0170ae000, 0xcb7c6, 0xf9f82, 0x0, 0x0)
        github.com/peterbourgon/diskv@v2.0.1+incompatible/diskv.go:550 +0xdc
github.com/peterbourgon/diskv.(*siphon).Read(0xc014d6b170,
0xc0019f0000, 0x1000, 0x1000, 0x778, 0x0, 0x0)
        github.com/peterbourgon/diskv@v2.0.1+incompatible/diskv.go:386 +0x12f
bufio.(*Reader).fill(0xc01cf6eea0)
        bufio/bufio.go:100 +0x103
bufio.(*Reader).ReadSlice(0xc01cf6eea0, 0x30000000000a, 0x158, 0x160,
0xc01474d080, 0x0, 0x0)
        bufio/bufio.go:359 +0x3d
encoding/csv.(*Reader).readLine(0xc000d09c38, 0x16, 0x8, 0xc01474d080,
0xc01458db00, 0x203000)
        encoding/csv/reader.go:218 +0x49
encoding/csv.(*Reader).readRecord(0xc000d09c38, 0x0, 0x0, 0x0, 0x1,
0xc01fecce18, 0x4, 0xc01fecce22, 0x7)
```

This was running on v2.0.1 but I don't see any substantive changes between then and HEAD in terms of the cache.

--- a/diskv.go
+++ b/diskv.go
@@ -630,6 +630,9 @@ func (d *Diskv) completeFilename(pathKey *PathKey) string {
 // cacheWithLock attempts to cache the given key-value pair in the store's
 // cache. It can fail if the value is larger than the cache's maximum size.
 func (d *Diskv) cacheWithLock(key string, val []byte) error {
+	// If the key already exists, delete it.
+	d.bustCacheWithLock(key)
+
 	valueSize := uint64(len(val))
 	if err := d.ensureCacheSpaceWithLock(valueSize); err != nil {
 		return fmt.Errorf("%s; not caching", err)

--- a/issues_test.go
+++ b/issues_test.go
@@ -3,6 +3,7 @@ package diskv
 import (
 	"bytes"
 	"io/ioutil"
+	"math/rand"
 	"sync"
 	"testing"
 	"time"
@@ -117,4 +118,46 @@ func TestIssue17(t *testing.T) {
 	}
 	close(start)
 	wg.Wait()
+}
+
+func TestIssue40(t *testing.T) {
+	var (
+		basePath = "test-data"
+	)
+	// Simplest transform function: put all the data files into the base dir.
+	flatTransform := func(s string) []string { return []string{} }
+
+	// Initialize a new diskv store, rooted at "my-data-dir", with a 1MB cache.
+	d := New(Options{
+		BasePath:     basePath,
+		Transform:    flatTransform,
+		CacheSizeMax: 100,
+	})
+
+	defer d.EraseAll()
+
+	// Write a 50 byte value
+	k1 := "key1"
+	d1 := make([]byte, 50)
+	rand.Read(d1)
+	d.Write(k1, d1)
+	// Get two read streams on it
+	s1, err := d.ReadStream(k1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	s2, err := d.ReadStream(k1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ioutil.ReadAll(s1)
+	ioutil.ReadAll(s2)
+
+	// Now write a different value
+	k2 := "key2"
+	d2 := make([]byte, 60)
+	rand.Read(d2)
+	d.Write(k2, d2)
+	// ... and read it
+	d.Read(k2)
 }

--- a/issues_test.go
+++ b/issues_test.go
@@ -120,6 +120,9 @@ func TestIssue17(t *testing.T) {
 	wg.Wait()
 }
 
+// Test for issue #40, where acquiring two stream readers on the same k/v pair
+// caused the value to be written into the cache twice, messing up the
+// size calculations.
 func TestIssue40(t *testing.T) {
 	var (
 		basePath = "test-data"
@@ -127,7 +130,8 @@ func TestIssue40(t *testing.T) {
 	// Simplest transform function: put all the data files into the base dir.
 	flatTransform := func(s string) []string { return []string{} }
 
-	// Initialize a new diskv store, rooted at "my-data-dir", with a 1MB cache.
+	// Initialize a new diskv store, rooted at "my-data-dir",
+	// with a 100 byte cache.
 	d := New(Options{
 		BasePath:     basePath,
 		Transform:    flatTransform,
@@ -136,12 +140,16 @@ func TestIssue40(t *testing.T) {
 
 	defer d.EraseAll()
 
-	// Write a 50 byte value
+	// Write a 50 byte value, filling the cache half-way
 	k1 := "key1"
 	d1 := make([]byte, 50)
 	rand.Read(d1)
 	d.Write(k1, d1)
-	// Get two read streams on it
+
+	// Get *two* read streams on it. Because the key is not yet in the cache,
+	// and will not be in the cache until a stream is fully read, both
+	// readers use the 'siphon' object, which always writes to the cache
+	// after reading.
 	s1, err := d.ReadStream(k1, false)
 	if err != nil {
 		t.Fatal(err)
@@ -150,14 +158,34 @@ func TestIssue40(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	// When each stream is drained, the underlying siphon will write
+	// the value into the cache's map and increment the cache size.
+	// This means we will have 1 entry in the cache map
+	// ("key1" mapping to a 50 byte slice) but the cache size will be 100,
+	// because the buggy code does not check if an entry already exists
+	// in the map.
+	// s1 drains:
+	//   cache[k] = v
+	//   cacheSize += len(v)
+	// s2 drains:
+	//   cache[k] = v /* overwrites existing */
+	//   cacheSize += len(v) /* blindly adds to the cache size */
 	ioutil.ReadAll(s1)
 	ioutil.ReadAll(s2)
 
-	// Now write a different value
+	// Now write a different k/v pair, with a 60 byte array.
 	k2 := "key2"
 	d2 := make([]byte, 60)
 	rand.Read(d2)
 	d.Write(k2, d2)
-	// ... and read it
+	// The act of reading the k/v pair back out causes it to be cached.
+	// Because the cache is only 100 bytes, it needs to delete existing
+	// entries to make room.
+	// If the cache is buggy, it will delete the single 50-byte entry
+	// from the cache map & decrement cacheSize by 50... but because
+	// cacheSize was improperly incremented twice earlier, this will
+	// leave us with no entries in the cacheMap but with cacheSize==50.
+	// Since CacheSizeMax-cacheSize (100-50) is less than 60, there
+	// is no room in the cache for this entry and it panics.
 	d.Read(k2)
 }

